src/main/java/com/conveyal/analysis/controllers/GtfsVectorTileMaker.java

```diff
@@ -0,0 +1,293 @@
+    private STRtree buildStopsIndex (String bundleScopedId) {
+        final long startTimeMs = System.currentTimeMillis();
+        GTFSFeed feed = gtfsCache.get(bundleScopedId);
+        STRtree stopsIndex = new STRtree();
+        LOG.info("{}: indexing {} stops", bundleScopedId, feed.stops.size());
+        for (Stop stop : feed.stops.values()) {
+            // This is inefficient, TODO specialized spatial index to bin points into mercator tiles (like hashgrid).
+            Envelope stopEnvelope = new Envelope(stop.stop_lon, stop.stop_lon, stop.stop_lat, stop.stop_lat);
+            stopsIndex.insert(stopEnvelope, stop);
+        }
+        stopsIndex.build();
+        LOG.info("Creating vector tile spatial index for stops in feed {} ({} ms)", bundleScopedId, System.currentTimeMillis() - startTimeMs);
+        return stopsIndex;
+    }
+
+    /** RemovalListener triggered when a spatial index is evicted from the cache. */
+    private void logCacheEviction (String feedId, STRtree value, RemovalCause cause) {
+        LOG.info("Vector tile spatial index removed. Feed {}, cause {}.", feedId, cause);
+    }
+
+    /** Produce a single Mapbox vector tile for the given unique GTFS feed ID and tile coordinates. */
+    public byte[] getTile (String bundleScopedFeedId, int zTile, int xTile, int yTile) {
+
+        // It might make sense to unify these two values into a single compound object.
+        STRtree shapesIndex = shapeIndexCache.get(bundleScopedFeedId);
+        STRtree stopsIndex = stopIndexCache.get(bundleScopedFeedId);
+
+        final long startTimeMs = System.currentTimeMillis();
+        final int tileExtent = 4096; // Standard is 4096, smaller can in theory make tiles more compact
+        Envelope wgsEnvelope = MapTile.wgsEnvelope(zTile, xTile, yTile);
+        Collection<Geometry> patternGeoms = new ArrayList<>(64);
+        for (LineString wgsGeometry : (List<LineString>) shapesIndex.query(wgsEnvelope)) {
+            Geometry tileGeometry = clipScaleAndSimplify(wgsGeometry, wgsEnvelope, tileExtent);
+            if (tileGeometry != null) {
+                patternGeoms.add(tileGeometry);
+            }
+        }
+        JtsLayer patternLayer = new JtsLayer("conveyal:gtfs:patternShapes", patternGeoms, tileExtent);
+
+        Collection<Geometry> stopGeoms = new ArrayList<>(64);
+        for (Stop stop : ((List<Stop>) stopsIndex.query(wgsEnvelope))) {
+            Coordinate wgsStopCoord = new Coordinate(stop.stop_lon, stop.stop_lat);
+            if (!wgsEnvelope.contains(wgsStopCoord)) {
+                continue;
+            }
+
+            Map<String, Object> properties = new HashMap<>();
+            properties.put("feedId", stop.feed_id);
+            properties.put("id", stop.stop_id);
+            properties.put("name", stop.stop_name);
+            properties.put("lat", stop.stop_lat);
+            properties.put("lon", stop.stop_lon);
+
+            // Factor out this duplicate code from clipScaleAndSimplify
```

Comment:
Should we factor this out?
```java
    private static Coordinate projectCoordinate (Coordinate c, Envelope envelope) {
        double px = ((c.x - envelope.getMinX()) * TILE_EXTENT) / envelope.getWidth();
        double py = ((envelope.getMaxY() - c.y) * TILE_EXTENT) / envelope.getHeight();
        return new Coordinate(px, py);
    }
```

Reply:
Yes, this can be factored out. I think the original code that clones a coordinate and sets its fields was the result of a series of refactors, starting from a version that tried to avoid new object creation. In the end we create a lot of small objects in the vector tile generation, which is not a big deal for contemporary garbage collectors.